### ISSUE TITLE
Use Memory Segment API for aligned vector loads. 

### ISSUE THREAD

**Issue by jatin-bhateja** [jbellis/jvector#132]
Hi All,

Most of the vectorized code in [SimdOps.java ](https://github.com/jbellis/jvector/blob/main/jvector-twenty/src/main/java/io/github/jbellis/jvector/vector/SimdOps.java) is using fromArray API to load the contents into vector.

With JDK-20+ Vector API added the support for loading and storing vectors from [MemorySegments](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/foreign/MemorySegment.html).

Using [from/intoMemorySegment APIs](https://docs.oracle.com/en/java/javase/21/docs/api/jdk.incubator.vector/jdk/incubator/vector/FloatVector.html#fromMemorySegment(jdk.incubator.vector.VectorSpecies,java.lang.foreign.MemorySegment,long,java.nio.ByteOrder)) one can ensures aligned vector load / store, given that most of the code is using SPECIES_PREFERRED which means the vector size (64 bytes) will match with the cacheline size on X86 AVX-512 targets. 

Thus if the first vector load in the vector loop happens from an address which is not a multiple of cacheline / vector size then each successive vector load will span across the cache line, this may have significant performance penalty. 

Following PMU events can be use to count the the number of SPLIT loads against total number of memory loads.

![image](https://github.com/jbellis/jvector/assets/59989778/72e5f9e3-e900-491e-b415-2138f07cdb03)

Best Regards,
Jatin

**Comment by tjake** [comment]
I did, @jatin-bhateja look at https://github.com/jbellis/jvector/pull/90 you can run the JMH yourself.

**Comment by jatin-bhateja** [comment]
Thanks for the link @tjake , I will take a look back.

**Comment by jatin-bhateja** [comment]
> Thanks for the link @tjake , I will take a look and get back.

Hi @tjake ,

I ran SimilarityBench JMH micro included with PR #90  after modifying the SPECIES to FloatVector.SPECIES_PREFERRED on Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz Icelake Server.

![image](https://github.com/jbellis/jvector/assets/59989778/6896f860-e4cf-4260-8ed5-f590fcc4daeb)

![image](https://github.com/jbellis/jvector/assets/59989778/57eb3b3b-becb-45ee-bc66-cd45ec374829)

Benchmarking was done over unmodified jvector-80-native-vectors branch after some minor build fixes. As can be seen with array based backing storage around 22% of vector loads are split across the cache lines.  This score improves with memory segments are we see around 5.12% split loads.

But what's bothering me more is that we see almost 2x number of loads with Memory Segment compared to array based backing storage.

Will get back with more analysis on this.

Best Regards,
Jatin

**Comment by jatin-bhateja** [comment]
> Thanks for the link @tjake , I will take a look and get back.

Hi @tjake ,

I ran SimilarityBench JMH micro included with PR #90  with following modifications on Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz Icelake Server.

```
IPROMPT>diff src/main/java/org/sample/SimilarityBench.java /home/jatinbha/sandboxes/jvector/jvector-native/src/test/java/microbench/SimilarityBench.java 
28d27
< import java.lang.foreign.Arena;
39,40c38,39
<     private static final int SIZE = 1024;
<     private static final VectorSpecies<Float> SPECIES = FloatVector.SPECIES_512;
---
>     private static final int SIZE = 2;
>     private static final VectorSpecies<Float> SPECIES = FloatVector.SPECIES_64;
43,44c42,43
<         MemorySegment fm0 = Arena.global().allocate(SIZE * Float.BYTES, 64);
<         MemorySegment fm1 = Arena.global().allocate(SIZE * Float.BYTES, 64);
---
>         MemorySegment fm0 = MemorySegment.ofBuffer(ByteBuffer.allocateDirect(SIZE * Float.BYTES).order(ByteOrder.LITTLE_ENDIAN));
>         MemorySegment fm1 = MemorySegment.ofBuffer(ByteBuffer.allocateDirect(SIZE * Float.BYTES).order(ByteOrder.LITTLE_ENDIAN));
85c84
<         for (int i = 0; i < SIZE * Float.BYTES; i += SPECIES.length() * Float.BYTES) {
---
>         for (int i = 0; i < SIZE; i += SPECIES.length()) {
109d107
< 


```

Following are the results along with relevant PMC counters.

![image](https://github.com/jbellis/jvector/assets/59989778/2a681487-44c8-419b-a38e-6b0d19a7691b)

![image](https://github.com/jbellis/jvector/assets/59989778/d99e25ad-8a1f-41b6-85b8-c7d895a25aed)



Benchmarking was done over unmodified jvector-80-native-vectors branch after some minor build fixes. As can be seen with array based backing storage around 78% of vector loads are split across the cache lines.  Split penalty significantly improves with memory segments as we see almost negligible split loads compared to total number of loads.  There is around 15% improvement in throughput.

Will spend more time to analyze NativeVectorizationProvider.

Best Regards,
Jatin

**Comment by tjake** [comment]
Hey @jatin-bhateja thanks for taking a look!

So looks like the ValueLayout isn't aligned and allocateDirect is? Am I reading it right?

**Comment by jatin-bhateja** [comment]
> Hey @jatin-bhateja thanks for taking a look!
> 
> So looks like the ValueLayout isn't aligned and allocateDirect is? Am I reading it right?

Yes,  JDK 21 introduced a new API [Arean.allocate](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/foreign/Arena.html#allocate(long,long)) to allocate aligned memory segments.

**Comment by tjake** [comment]
Hi @jatin-bhateja I was able to reproduce the split-load drop with aligned memory but I don't see a 15% bump.  I only see a ~5% improvement over arrays.  Any idea why?

Also, since these are vector embedings mostly 1024 is pretty large.  When I run with 128 floats I see a 2% loss over arrays.  With 1536 (openai embedding size) I see 11% improvement.

**Comment by jatin-bhateja** [comment]
> When I run with 128 floats I see a 2% loss over arrays

Hi @tjake 
It all depends on the %age of cycles spent in computation loop compared to rest of the application, small number of iterations may not impact overall throughput significantly,  which goes in line with your observations.

**Comment by r-devulap** [comment]
AFAIK, JVector now uses MemorySegments for all the native code. Closing this, please let me know if you disagree.
